Public/Src/Cache/ContentStore/Library/Stores/FileSystemContentStoreInternal.cs

```diff
@@ -2075,18 +2075,18 @@ public Task<EvictResult> EvictAsync(Context context, ContentHashWithLastAccessTi
         }
 
         /// <inheritdoc />
-        public async Task<DeleteResult> DeleteAsync(Context context, ContentHash contentHash)
+        public async Task<DeleteResult> DeleteAsync(Context context, ContentHash contentHash, bool allowWait = true)
         {
-            var evictResult = await EvictCoreAsync(context, new ContentHashWithLastAccessTimeAndReplicaCount(contentHash, DateTime.MinValue, safeToEvict: true), force: true, onlyUnlinked: false, (l) => { });
+            var evictResult = await EvictCoreAsync(context, new ContentHashWithLastAccessTimeAndReplicaCount(contentHash, DateTime.MinValue, safeToEvict: true), force: true, onlyUnlinked: false, (l) => { }, ensureSuccess: allowWait);
             return evictResult.ToDeleteResult(contentHash);
         }
 
-        private async Task<EvictResult> EvictCoreAsync(Context context, ContentHashWithLastAccessTimeAndReplicaCount contentHashInfo, bool force, bool onlyUnlinked, Action<long> evicted)
+        private async Task<EvictResult> EvictCoreAsync(Context context, ContentHashWithLastAccessTimeAndReplicaCount contentHashInfo, bool force, bool onlyUnlinked, Action<long> evicted, bool ensureSuccess = false)
         {
             ContentHash contentHash = contentHashInfo.ContentHash;
 
             long pinnedSize = 0;
-            using (LockSet<ContentHash>.LockHandle? contentHashHandle = _lockSet.TryAcquire(contentHash))
+            using (LockSet<ContentHash>.LockHandle? contentHashHandle = ensureSuccess ? await _lockSet.AcquireAsync(contentHash) : _lockSet.TryAcquire(contentHash))

```

Comment:
I'm not sure about the name `ensureSuccess` because it is not quite the case here. The flag just blocks the operaiton.

So maybe: `acquireLock`? or something?

Reply:
I mean, sure. Works for me!